File: 序号/23.合并K个升序链表.cpp
class Solution {
  struct Cmp {
    bool operator()(ListNode* a, ListNode* b) { return a->val > b->val; }
  };

 public:
  /** 每个链表都是升序，合并成一个升序链表 */
  ListNode* mergeKLists(vector<ListNode*>& lists) {
    // 思路1 有限队列
    priority_queue<ListNode*, vector<ListNode*>, Cmp> q;
    for (ListNode* node : lists) {
      while (node) {
        q.emplace(node);
        node = node->next;
      }
    }

    ListNode dummy;
    ListNode* p = &dummy;
    while (!q.empty()) {
      auto top = q.top();
      p->next = top;
      q.pop();
      p = p->next;
    }
    p->next = nullptr;
    return dummy.next;
  }
};

File: 序号/23. 合并 K 个升序链表.py
ListNode.__lt__=lambda a,b:a.val<b.val
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        d=[]
        for h in lists:
            if h:
                heapq.heappush(d,h)
        p=fake = ListNode(0)
        while d:
            top = heapq.heappop(d)
            p.next=top
            p=top
            if top.next:
                heapq.heappush(d,top.next)
        return fake.next